ide/posts/2020-06-50-ghcide-and-nixpkgs.md

```diff
@@ -0,0 +1,321 @@
+---
+title: Getting ghcide into nixpkgs
+author: Malte Brandy
+date: 2020-06-05
+github: maralorn
+---
+
+This is a guest entry by maralorn an enthusiastic ghcide user.
+
+A few weeks ago I got ghcide into nixpkgs, the package set of the package
+manager nix and the distribution nixos. Mind you, that was not a brave act of
+heroism or dark wizardry. Once I grasped the structure of the nixpkgs Haskell
+ecosystem, it was actually pretty easy. In this post I wanna share my
+experience and tell you what I learned about the nixpkgs Haskell infrastructure
+and ghcide.
+It took me quite a while to collect the knowledge lain out below, so I gather
+it might be useful for someone else. But feel free to skip the parts that
+aren‘t of interest to you. This is a very explanatory post which is probably
+already much too long so I won‘t go to much into the details of actual nix
+code. If you just want to know how to use ghcide in a nix environment jump to
+the end of the post or just consult the README of ghcide.
+
+<!--more-->
+
+## What’s the problem?
+
+Haskell development tooling setup is infamous for being brittle and hard to
+setup. Every other day when someone asks on reddit or in the `#haskell`
+channel, inescapably there will come at least one answer of the form "It’s not
+worth the pain. Just use ghcid." I guess one point of this blog series is that
+this does not have to be the case anymore.
+
+But what were the reasons for this resignation? One is certainly that `ghcid`
+is a really great and easy to use tool. But I think it‘s clear that a well done
+language server can leverage you much further and to me `ghcide` has already
+proven this.
+
+### Compile your project and ghcide with the same ghc!
+
+One source of frustration is likely that succesfully setting up a language
+server that is deeply interwoven with `ghc` like `ghcide` has one very
+important requirement. **You need to compile `ghcide` with the same `ghc`
+(version) as your project.** This shouldn‘t be hard to achieve nowadays - I’ll
+show how to do it if you use `nix` in this blogpost and I assume it‘s the
+default in other setups - but if you fail to meet this requirement you are in
+for a lot of trouble.
+
+So why exactly do we need to use "the same ghc" and what does that even mean?
+Frankly I am not totally sure. I am not a `ghcide` developer. I guess sometimes
+you can get away with some slight deviations. But the general recommendation is
+to use the same ghc version. I can tell you three situations that will cause
+problems or have caused problems for me:
+
+1. Using another `ghc` release. E.g. using `ghcide` compiled with `ghc` 8.6 on
```

Comment:
Can you include some example errors of these three cases? So that if someone runs into them then they get pointed to this blog posts.

Reply:
I can try.

Reply:
I have found error messages for everything but the patch problem. But I guess that one would depend very much on the patch anyways.